### PR TITLE
[Do not merge] Adding run phase1/2 for each provider

### PR DESCRIPTION
**Pull request by cdrage**
WIP. replaces #479 pr.

Todo:

- [x] Add Phase1 and Phase2 run to Kubernetes run
- [x] Add Phase1 and Phase2 run to Docker run
- [x] Add Phase1 and Phase2 run to OpenShift run
- [ ] Add Phase1 and Phase2 run to Mesos run

...

| Provider      | Implementation |
| ------------- | -------------- |
| Docker        | __Phase 1:__ Do nothing. Possibly check to see if containers are already ran? __Phase 2:__ Run requested containers on local machine. |
| Kubernetes    | __Phase 1:__ Initially run with replication set to 0 to test deployment. __Phase 2:__ Increase replica count from 0 to >0. |
| Openshift     | __Phase 1:__ Initially run with replication set to 0 to test deployment. __Phase 2:__ Increase replica count from 0 to >0, OR initiate template. |

### CODE REVIEW

File: atomicapp/providers/docker.py [hunk 1]
```diff
@@ -77,12 +79,23 @@ def _get_containers(self):
         else:
             return dict((line, 1) for line in subprocess.check_output(docker_cmd, shell=True).splitlines())
 
-    def deploy(self):
-        logger.info("Deploying to provider: Docker")
+    def _run_phase1(self):
```
Comment (rtnpro) [inline atomicapp/providers/docker.py]: Not a great name :P

File: atomicapp/providers/docker.py [hunk 2]
```diff
@@ -77,12 +79,23 @@ def _get_containers(self):
...
+        logger.info("Initializing Docker container")
+        logger.debug("Phase 1 run: Docker")
         for container in self._get_containers():
             if re.match("%s_+%s+_+[a-zA-Z0-9]{12}" % (self.namespace, self.image), container):
-                raise ProviderFailedException("Container with name %s-%s already deployed in Docker" % (self.namespace, self.image))
-
+                raise ProviderFailedException(
+                    "Container with name %s-%s already deployed in Docker" % (self.namespace, self.image))
+
+    def _run_phase2(self):
```
Comment (rtnpro) [inline atomicapp/providers/docker.py]: It could be ``_run``

File: atomicapp/providers/kubernetes.py [hunk 3]
```diff
@@ -167,24 +170,92 @@ def _scale_replicas(self, path, replicas=0):
...
                 cmd.append("--kubeconfig=%s" % self.config_file)
             self._call(cmd)
 
-    def undeploy(self):
+    def _run_phase2(self):
+        """Run Phase 2
+        If a Pod is detected, we run it
+        If a Replication Controller is detected, we scale to N in the artifact
+        All other artifacts are ignored as they are presumed to be ran during the
+        installation phase.
```
Comment (kadel) [inline atomicapp/providers/kubernetes.py]: 'installation phase'? Shouldn't this say 'during phase 1'?

**Comment by cdrage** [comment]
Don't merge until post-GA.

I'm going to close this for now but keep the branch open for future reference

**Pull request closed without merging**